File: internal/filter/filter.go
package filter

import "regexp"

type Filter struct {
	allow []*regexp.Regexp
	deny  []*regexp.Regexp
}

func New(allowPatterns []string, denyPatterns []string) *Filter {
	f := &Filter{allow: compile(allowPatterns), deny: compile(denyPatterns)}
	return f
}

func compile(patterns []string) []*regexp.Regexp {
	out := make([]*regexp.Regexp, 0, len(patterns))
	for _, p := range patterns {
		r, err := regexp.Compile("(?i)" + p)
		if err == nil {
			out = append(out, r)
		}
	}
	return out
}

func (f *Filter) Matches(command string) bool {
	if command == "" {
		return false
	}
	allowed := len(f.allow) == 0
	for _, r := range f.allow {
		if r.MatchString(command) {
			allowed = true
			break
		}
	}
	if !allowed {
		return false
	}
	for _, r := range f.deny {
		if r.MatchString(command) {
			return false
		}
	}
	return true
}

File: src/filter.ts
interface FilterConfig {
  allow?: string[];
  deny?: string[];
}

function compile(patterns: string[]): RegExp[] {
  return patterns
    .map((pattern) => {
      try {
        return new RegExp(pattern, "i");
      } catch {
        return null;
      }
    })
    .filter((r): r is RegExp => Boolean(r));
}

export function createFilter(filterConfig: FilterConfig) {
  const allowList = compile(filterConfig.allow || []);
  const denyList = compile(filterConfig.deny || []);

  return {
    matches(command: string): boolean {
      const normalized = `${command || ""}`.trim();
      if (!normalized) {
        return false;
      }

      const allowed = allowList.length === 0 || allowList.some((regex) => regex.test(normalized));
      const denied = denyList.some((regex) => regex.test(normalized));
      return allowed && !denied;
    }
  };
}
